### ISSUE TITLE
Defining AbstractModuleSource interface type

### ISSUE THREAD

**Issue by guybedford** [whatwg/webidl#1322]
The [Source Phase Imports](https://github.com/tc39/proposal-source-phase-imports) proposal requires that objects returned for the source phase must inherit from the new `AbstractModuleSource` intrinsic in their prototype chain.

In order for the WebAssembly ESM Integration to express this inheritance in WebIDL, once Stage 3 is obtained for the proposal, I believe `AbstractModuleSource` would need to be defined as a new interface type.

Creating this as a new tracking issue for this interface type definition.

**Comment by bathos** [comment]
This seems akin to how a DOMException interface object inherits from %Error%. I don’t see how defining an AbstractModuleSource Web IDL interface would help with it — that would just create a collision?

**Comment by guybedford** [comment]
Good points. We probably should support both `%AbstractModuleSource% and `%AbstractModuleSource.prototype%` chains here yes.

Perhaps that would mean two new attributes then?

```
[Exposed=*, Serializable, InstrinsicParent="%Error%", IntrinsicPrototypeParent="%Error.prototype%"]
interface DOMException { /*...*/ };

[LegacyNamespace=WebAssembly, Exposed=(Window,Worker,Worklet), IntrinsicParent="%AbstractModuleSource%", IntrinsicPrototypeParent="%AbstractModuleSource.prototype%"]
interface Module { /*...*/ };
```

It's quite a lot of machinery for a single case though that only possible simplifies DOMException. Have there been other cases of this sort of thing?

While DOMException is an exception, we do have a similar thing for `namespaces` in the console spec at https://console.spec.whatwg.org/#console-namespace where it overrides the default prototype from the https://webidl.spec.whatwg.org/#namespace-object steps effectively with a comment. Perhaps we could use a similar approach.

**Comment by annevk** [comment]
I wonder if @yuki3 or @saschanaz have thoughts about this.

**Comment by yuki3** [comment]
My very personal preference is to define things systematically rather than defining things with ad-hoc rules (so, I tend to like the idea of [IntrinsicHeritage]). Having said that...

1. Once it's explicitly supported (showing the existence of [IntrinsicHeritage] explicitly), I'm afraid that people would start abusing it more and more. E.g. "Oh, we can use this feature for our own unique purpose! (although not in an originally designed way)". I feel it's hard to prevent people from using existing features even when the features are explicitly marked as obsolete, deprecated, unrecommended, unsafe, etc.
2. Given we have only two cases, I don't see the strong necessity of [IntrinsicHeritage].
3. I'm fine with a special requirement of HTMLAllCollection's [[Call]] being not supported by an extended attribute.
4. I'm fine with a special requirement of Location's "valueOf" property being not supported by an extended attribute.
5. ...

I understand I'm talking about a sort of ...politics?... and not design of software, architecture, etc. Purely technically talking, [IntrinsicHeritage] looks good. I understand %X.prototype% is different from %X%.prototype, but [IntrinsicPrototypeParent] looks a bit overkilling/redundant (we don't want to support totally different inheritances for interface objects and prototype objects but it looks possible). Can we implicitly assume %X.prototype% inheritance, too?

In summary, my preference is: 1. special rule (no extended attribute) is enough (and good in total imho), 2. either of [IntrinsicHeritage] or [IntrinsicParent] + [IntrinsicPrototypeParent].

**Comment by petervanderbeken** [comment]
1. Given how few cases there are at this point I also don't think we should add an extended attribute for this just yet.
2. If this is done through an extended attribute there should only be one which then changes the prototype chain for both interface and interface object at the same time. For `DOMException` that would then fix #1107 too.

**Comment by bathos** [comment]
Declarative comment / prose text “overrides” both have an unclear relationship to the imperative realm-initializing algorithms in Web IDL (_when_ do the “overrides” happen?) and make it impossible for tooling that consumes Web IDL fragments to know about exceptional cases without hard-coding knowledge of the special cases. Since I work on that kind of tooling, I’m biased to prefer an extended attribute solution.

In the past I think some EAs that aren’t _quite_ “legacy” were given the `Legacy` prefix to discourage new “unbridled” usage?

If a non-IDL solution is used, then it would suggest to me that the existing imperative step for taking care of DOMException...

![3.7.3. Interface prototype object, “create an interface object,” step 4](https://github.com/whatwg/webidl/assets/6257356/e25c40c0-7d1e-40d8-be94-efd8ca1759c5)

...would want to be removed? Or is the idea that the “create an interface object” algorithm would introduce a new step like this for `WebAssembly.Module` even though the interface is defined elsewhere?

BTW I think there is another example of this that’s also in WebAssembly? The NativeError-template Error classes. I’m unsure if they would benefit, but they do use use %Error%/%Error.prototype%.

**Comment by guybedford** [comment]
Would it be possible to directly permit intrinsic inheritance for interfaces of the form:

```
interface DOMException : %Error% { ... }

interface Module : %AbstractModuleSource% { ... }
```

where this restricts intrinsic inheritance to be a specific case only - ie these must be well-defined intrinsic objects being subclassed, hopefully avoiding some kind of arbitrary object subclassing.

and then this intrinsic inheritance could itself be defined to include both the constructor and prototype inheritance chains, effectively in line with JS class inheritance.

**Comment by saschanaz** [comment]
Adding a new syntax is more expensive to the infra than the extended attribute option. Given that the latter is already being questioned as there are only few cases, I doubt we would go that path.

**Comment by bathos** [comment]
In addition to what @saschanaz said, new Web IDL syntax probably wouldn’t make sense here because this is really about the ES binding (which is what the extended attributes defined by Web IDL are for).

**Comment by guybedford** [comment]
Okay, so it sounds like if anything it would be a single attribute with some handling for class-like behaviours. If we restricted the attribute to only applying to JS intrinsics though would that mitigate concerns of overuse? Or would such an approach still need to be better motivated?

**Comment by domenic** [comment]
I think we've gotten a pretty clear signal from multiple implementers that prose-based overrides for these rare cases is preferred?

**Comment by bathos** [comment]
@domenic It does look like that — the two people who were asked earlier in the thread both stated that preference, and I’m not an implementor (at least not in a sense that I would expect to count towards such a decision).

**Comment by yuki3** [comment]
About when prose text "overrides" happens, I think it's nice to have something like "run an interface-specific configuration algorithm" in the imperative realm-initializing algorithms in Web IDL, and let each spec define "run an interface-specific configuration algorithm". This way has already been used in several places, right?

Just as a random idea; It might make sense and be helpful to have an extended attribute like [InterfaceSpecificConfiguration] in order to indicate that the interface has its own unique prose text "overrides". For example, we can use the extended attribute to cross-check whether we appropriately hard-code the behavior.

I'm skeptical about getting rid of hard-coded behaviors even if [IntrinsicHeritage] were introduced (e.g. Location's valueOf, HTMLAllCollection's [[Call]]). I don't think it will be helpful in case of Chromium. However, I don't have a strong objection especially when it will be very helpful for other people and/or projects. The concern about [IntrinsicParent] + [IntrinsicPrototypeParent] will remain, though.
